### PR TITLE
refactor: gas settings

### PR DESCRIPTION
**Pull request by guanbinrui**
## Description

Closes # (NO_ISSUE)

### Component Hierarchy Tree

```tsx
<SettingsContext>
    <SettingsBoard>
        <GasSection />
        <SlippageToleranceSection />
    </SettingsBoard>
</SettingsContext>
```

### API

```ts
const selectAdvancedSettings = useSelectAdvancedSettings({
    pluginID: NetworkPluginID.PLUGIN_EVM,
    chainId: ChainId.Mainnet,
})

const onSelect = useCallback(async () => {
    const settings = await selectAdvancedSettings()

    settings.slippageTolerance // number
    settings.gasOption // Web3Helper.GasOptionAll
})
```

## Type of change

- [ ] Documentation
- [x] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

### CODE REVIEW

File: packages/shared/src/UI/components/SettingsBoard/Context.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        value: gasOptions,
+        loading: gasOptionsLoading,
+        error: gasOptionsError,
+        retry: gasOptionRetry,
+    } = useGasOptions<'all'>()
+
+    // useEffectOnce(() => {
+    //     if (!gasOptions)
+    // }, [gasOptions])
+
```
Comment (UncleBill) [inline packages/shared/src/UI/components/SettingsBoard/Context.tsx]: Don't forget to remove.

File: packages/shared/src/UI/components/SettingsBoard/Section.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+export function Section(props: SectionProps) {
+    const { title, additions, children } = props
+    const { classes } = useStyles()
+
+    return (
+        <div className={classes.root}>
+            <Box display="flex" alignItems="center" justifyContent="space-between">
+                <Typography className={classes.title}>{title}</Typography>
+                {additions ?? <Typography> </Typography>}
```
Comment (UncleBill) [inline packages/shared/src/UI/components/SettingsBoard/Section.tsx]: `<Typography> </Typography>` An empty element?

File: packages/shared/src/UI/components/SettingsBoard/hooks/useCallbackOnce.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { DependencyList, useCallback, useRef } from 'react'
```
Comment (UncleBill) [inline packages/shared/src/UI/components/SettingsBoard/hooks/useCallbackOnce.ts]: Move this outside the context of `SettingsBoard`?

File: packages/shared/src/UI/components/SettingsBoard/hooks/useCallbackOnce.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+export function useCallbackOnce<T extends Function>(callback: T, deps: DependencyList) {
+    const flag = useRef<boolean>(false)
+    const callback_ = (...args: any[]) => {
+        if (flag.current) return
+        flag.current = true
+        callback(...args)
+    }
+
+    return useCallback(callback_ as unknown as T, deps.concat(flag))
```
Comment (UncleBill) [inline packages/shared/src/UI/components/SettingsBoard/hooks/useCallbackOnce.ts]: `flag` is a `ref`, no need to concat it.
```suggestion
    return useCallback(callback_ as unknown as T, deps)
```

File: packages/shared/src/contexts/UITaskManager.tsx [hunk 5]
```diff
@@ -4,23 +4,19 @@ import { EMPTY_LIST } from '@masknet/shared-base'
...
 }
 
 export interface BaseDialogProps<T> extends Pick<InjectedDialogProps, 'open' | 'onClose'> {
-    onSubmit?(result: T): void
+    onSubmit?(result: T | null): void
 }
 
 /**
  * Create a manager of small UI task sessions,
  * which provide both a Context and a Provider.
```
Comment (UncleBill) [inline packages/shared/src/contexts/UITaskManager.tsx]: A grammar error written by me.
```suggestion
 * which provides both a Context and a Provider.
```

File: packages/shared/src/contexts/UITaskManager.tsx [hunk 6]
```diff
@@ -68,16 +64,16 @@ export const createUITaskManager = <
                         Component,
                         {
                             key: task.id,
-                            open: true,
                             ...task.options,
+                            open: true,
```
Comment (UncleBill) [inline packages/shared/src/contexts/UITaskManager.tsx:68]: I think we should move `key: task.id` after `...task.options` as well.
```suggestion
                            open: true,
                            key: task.id,
```

File: packages/shared/src/UI/components/SettingsBoard/Context.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+export function useSettingsContext(initial?: {
+    pluginID?: NetworkPluginID
+    chainId?: Web3Helper.ChainIdAll
+    transaction?: Web3Helper.TransactionAll
+}) {
+    const t = useSharedI18N()
+    const pluginID = useCurrentWeb3NetworkPluginID(initial?.pluginID)
+    const [chainId] = useState(initial?.chainId)
+    const [transaction] = useState(initial?.transaction)
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/Context.tsx]: If you don't need to set these states, I don't think you need to `useState` here.

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+import { GasOptionType, NetworkPluginID } from '@masknet/web3-shared-base'
+import { useWeb3State } from '@masknet/plugin-infra/web3'
+import type { z as zod } from 'zod'
+import { zodResolver } from '@hookform/resolvers/zod'
+import { useGasSchema } from './hooks'
+
+const useStyles = makeStyles()((theme) => {
+    return {
+        unit: {
+            color: MaskColorVar.textLight,
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasForm.tsx]: Please don't use `MaskColorVar`. Synchronizing develop and use ` theme.palette.maskColor.input`.

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+            },
+            '& input[type=number]::-webkit-outer-spin-button': {
+                '-webkit-appearance': 'none',
+            },
+            '& input[type=number]::-webkit-inner-spin-button': {
+                '-webkit-appearance': 'none',
+            },
+        },
+        caption: {
+            color: MaskColorVar.textSecondary,
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasForm.tsx]: ditto, plz use the correct color.

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    const isEIP1559 = Others?.chainResolver.isSupport(chainId, 'EIP1559')
+
+    const schema = useGasSchema(chainId, transaction, gasOptions)
+
+    const methods = useForm<zod.infer<typeof schema>>({
+        shouldUnregister: false,
+        mode: 'onChange',
+        resolver: zodResolver(schema),
+        defaultValues: {
+            gasLimit: (transaction.gas as string | undefined) ?? '210000',
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasForm.tsx]: The default value should be 21000. Is this the default value for non-evm networks as well? If not I think we need to define a constant in a different network.

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        onChange?.(!error ? payload : undefined)
+    }, [error, isEIP1559, gasLimit, gasPrice, maxFeePerGas, maxPriorityFeePerGas])
+
+    return (
+        <FormProvider {...methods}>
+            <Grid container direction="row" spacing={2}>
+                <Grid item xs={isEIP1559 ? 12 : 6}>
+                    <Controller
+                        render={({ field }) => (
+                            <MaskTextField
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasForm.tsx:130]: I don't think we should use `MaskTextField` here, it should only be used in the dashboard. cc @Lanttcat

**Review by nuanyang233 — CHANGES_REQUESTED** [review]

File: packages/shared/src/UI/components/SettingsBoard/GasOptionSelector.tsx [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+import { GasOptionType, NetworkPluginID } from '@masknet/web3-shared-base'
+import { ChainId, formatGweiToWei, GasOption, Transaction } from '@masknet/web3-shared-evm'
+import { useWeb3State } from '@masknet/plugin-infra/web3'
+import { GasOption as GasOptionItem } from './GasOption'
+import { SettingsContext } from './Context'
+
+const useStyles = makeStyles()((theme) => {
+    return {
+        root: {
+            boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.05)',
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasOptionSelector.tsx]: Dark mode friendly?

File: packages/shared/src/UI/components/SettingsBoard/GasSection.tsx [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+enum GasSettingsType {
+    Basic = 'Basic',
+    Advanced = 'Advanced',
+}
+
+const useStyles = makeStyles()((theme) => {
+    return {
+        root: {},
+        paper: {
+            boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.05)',
```
Comment (nuanyang233) [inline packages/shared/src/UI/components/SettingsBoard/GasSection.tsx]: ditto

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx (hunk 10 above)
Comment (guanbinrui) [reply packages/shared/src/UI/components/SettingsBoard/GasForm.tsx]: It's an EVM-only form.

File: packages/shared/src/UI/components/SettingsBoard/GasForm.tsx (hunk 11 above)
Comment (guanbinrui) [reply packages/shared/src/UI/components/SettingsBoard/GasForm.tsx:130]: It's imported from `@masknet/theme`.

Comment (nuanyang233) [reply packages/shared/src/UI/components/SettingsBoard/GasForm.tsx:130]: Yes, but now it looks different from the color of the design.

Comment (guanbinrui) [reply packages/shared/src/UI/components/SettingsBoard/GasForm.tsx:130]: Really? But it looks the same as the Figma of this PR.

**Pull request merged**